### PR TITLE
Implement IUOs in terms of Optionals that are force-unwrapped when needed.

### PR DESCRIPTION
**Pull request by rudkx**
This set of commits moves us from treating IUOs as a separate type in the type system, instead treating them as Optionals that are force-unwrapped as needed based on setting up disjunctions in the ConstraintSystem that attempt to first type-check these as Optional, and only if that fails attempts to type-check as the underlying type.

### CODE REVIEW

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
FYI, I'm splitting my changes out into bite-sized chunks and will likely be pushing more of them here before merging. This is currently a whole bunch of NFC because it doesn't actually switch over from treating 'T!' as 'T?'.

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2331,6 +2331,44 @@ class ConstraintSystem {
...
+
+    Type underlyingType;
+
+    // FIXME: Support unwrapping results of function calls.
+    assert(!type->is<AnyFunctionType>());
+
+    underlyingType = type->getWithoutSpecifierType()->getAnyOptionalObjectType();
+    assert(underlyingType);
+
+    if (type->is<LValueType>())
```
Comment (slavapestov) [inline lib/Sema/ConstraintSystem.h]: Do both come up here? Maybe add comments explaining why/when

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: lib/Sema/ConstraintSystem.h (hunk 1 above)
Comment (rudkx) [reply lib/Sema/ConstraintSystem.h]: I suspect `InOut` shouldn't happen so I'll switch that to an assert in a separate commit.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
Yay! (?) The swiftpm crash appears to reproduce on macOS now.

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@swift-ci Please test

**Comment by rudkx** [comment]
@DougGregor @xedin @slavapestov 

I am still working through failures and know there is at least another decl checking issue and some PrintAsObjC and IDE printing issues, a couple source compatibility issues, and probably other issues left to resolve, but I think this is in pretty good shape to take a first look at. I can resolve any feedback in future commits.

**Comment by rudkx** [comment]
I've been holding off merging this because of a source compatibility suite failure due to not having enough changes pushed out yet, but I don't want this to keep hanging, so I've started splitting out the individual changes and merging them separately.

**Comment by rudkx** [comment]
I'm merging all of these as separate PRs so I no longer need this PR.

**Pull request closed without merging**